aesara/tensor/shape.py

```diff
@@ -397,169 +377,197 @@ class SpecifyShape(COp):
     -----
     Maybe in the future we will never do the assert!
 
-    We currently don't support specifying partial shape information.
-
-    TODO : test this op with sparse. Do C code for them too.
-
+    TODO: Implement rewrite to remove useless SpecifyShape when:
+        1. Input shape is already fully known
+        2. All unknown input shapes are associated with NoneConst
```

Comment:
Maybe this does not quite make sense... depends on whether Aesara complains if an intermediate variable of fixed shape is given a value with the wrong shape

Reply:
We can also narrow down the scope of such useless rewrite, to the cases where the not-none specified shape are constants, because then we can eagerly check if they match with the input type shape. This is also the most common use of SpecifyShape